src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -7347,15 +7347,22 @@ private StatementSyntax ParseEmbeddedStatement(bool complexCheck)
                 statement = SyntaxFactory.EmptyStatement(EatToken(SyntaxKind.SemicolonToken));
             }
 
-            // An "embedded" statement is simply a statement that is not a labelled
-            // statement or a declaration statement.
             switch (statement.Kind)
             {
+                // An "embedded" statement is simply a statement that is not a labelled
+                // statement or a declaration statement.
                 case SyntaxKind.LabeledStatement:
                 case SyntaxKind.LocalDeclarationStatement:
                 case SyntaxKind.LocalFunctionStatement:
                     statement = this.AddError(statement, ErrorCode.ERR_BadEmbeddedStmt);
                     break;
+                // Expression statements as embedded statements should be followed by semicolon.
+                case SyntaxKind.ExpressionStatement:
+                    if (((ExpressionStatementSyntax)statement).SemicolonToken.IsMissing)
```

Comment:
This will produce a duplicate diagnostic in both interactive and non-interactive scenarios when the compiler found a semicolon to be missing, whether it was an end-of-file or not.